feat(0020): add node rebalance EIP (#64)
emqx/eip: active/0020-node-evacuation.md

a particular node.

## Motivation

A user might want to shut down arbitrary nodes, e.g., to maintain the underlying
software or hardware. Or move connections from higher-loaded nodes to lower.

We need to minimize data loss when a node is switched off. The aim of node evacuation
is to minimize session data loss.

Since an MQTT client may choose to preserve its session between reconnects, we should
keep it, no matter whether the client is connected or not.
In EMQX 4.x, without an external database as the session storage extension (which is only available in the Enterprise edition)
the session is kept in the memory of the last node to which the client was connected.

So if a node shuts down, sessions of the clients currently or recently connected to it are lost.

Node evacuation should transfer the maximum possible amount of connections and sessions to the other nodes.

Node rebalancing should transfer a calculated (balanced among the nodes) number of connections and sessions from higher loaded nodes to other nodes.

For evacuation, after it's completed in one node, the node can be shut down by sysadmin (but not automatically).

> When the cluster reaches balance, all nodes should exit the connection prohibiting state for rebalancing.

Author: Tried to clarify  a bit: https://github.com/emqx/eip/pull/64/commits/237a99857ea836e9c2fc34ef172e125d8b1e528f
Is the description clearer now? 😅

This label is used to avoid a clash between different Rebalance Coordinators, like evacuation and rebalancing.

#### Internal API

Eviction Agent exposes API for the following actions:
* In disabled state:
    * enable the agent;
* In either of enabled states:
    * disable the agent;
    * provide stats about connected/disconnected client count;
    * provide stats about session count;
    * disconnect certain amount of connected clients;
    * evict certain amount of sessions.

#### HTTP API

Eviction Agent exposes two endpoints: `/node_eviction/status` and `/node_eviction/heath_check`.

#### Algorithm

##### `disabled` state

The disabled state is the default. In the `disabled` state Eviction Agent does not influence the node anyhow.

HTTP responses are the following:

Request:
```
GET /api/v4/node_eviction/status
```

Response:
```
HTTP/1.1 200 OK

{"status": "disabled"}
```

Request:
```
GET /api/v4/node_eviction/heath_check
```

Response:
```
HTTP/1.1 200 OK
```

The internal API behavior is the following:

```erlang
disabled = emqx_eviction_agent:status().
```

##### Enable

Enabling Eviction Agent requires two arguments:

> * `kind` label.

Reviewer: Q: can it be anything, or is it an enumeration of choices?

Author: Any atom that makes sence only for the client of the API.

* Server reference for [Server redirect](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255)

When being enabled, Eviction Agent enables prohibiting hooks: `client.connect` and `client.connack`;

The internal API is the following:

```erlang
ok = emqx_eviction_agent:enable(node_evacuation, <<"host1:port1 host2:port2">>).
```

##### Enabled (`enabled`)

When enabled, Eviction Agent has two working hooks. Their logic is:
* `client.connect` hook returns `CONNACK_SERVER` (for MQTT 3) or `RC_USE_ANOTHER_SERVER` (for MQTT 5) error.
* `client.connack` sets `'Server-Reference'` property to the specified servers `host1:port1 host2:port2 ...` for MQTT 5
clients acked with `RC_USE_ANOTHER_SERVER`. `'Server-Reference'` property is not set if servers are not specified or
for MQTT 3 clients.

HTTP responses are the following:

Request:
```
GET /api/v4/node_eviction/status
```

Response:
```
HTTP/1.1 200 OK

{
    "status": "enabled",
    "stats": {
        "current_connected": 25,
        "current_sessions": 50
    }
}
```

Request:
```
GET /api/v4/node_eviction/heath_check
```

Response:
```

> HTTP/1.1 503 OK

Reviewer: ```suggestion
HTTP/1.1 503 Service Unavailable
```
?

so it restores its state even after node restart.

Its logic is relatively straightforward.

#### Algorithm

##### Starting

At the start, the Evacuation Coordinator reads the settings from the local file.
* If evacuation is enabled Evacuation Coordinator tries to pass to the `evicting` state. It enables the local Eviction Agent.
    * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passed to the `evicting` state.
    * If the local Eviction Agent is already enabled with a different `kind` (i.e., is occupied by some other coordinator, like rebalancing), then the Evacuation Coordinator passes to the `disabled` state.
* The Evacuation Coordinator passes to the `disabled` state if evacuation is not enabled.

##### `disabled` state
This is the default no-op state.

##### Enabling
On enabling, Evacuation Coordinator:
* preserves settings in the local file;
* evacuation Coordinator enables local Eviction Agent;
* passes to the evicting state.

The internal API is:
```erlang
ok = emqx_node_rebalance_evacuation:start(
    #{
        conn_evict_rate => 10,

>         server_reference => <<"srv:123">>

Reviewer: Q: since the server reference can be a list of servers, maybe this could be a list as well?

Author: MQTT doesn't specify exact Server-Reference format, so finally, I decided not to complexify things and treat it as an opaque value.

Reviewer: From the spec, I think that it should be a space-separated list of servers.  Each server format is unspecified, but space-separated: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901255

> The Server Reference is a UTF-8 Encoded String. The value of this string is a space separated list of references. The format of references is not specified here.

    }
).
```

##### `evicting` state
In evicting state Evacuation Coordinator with the help of Eviction Agent:
 * constantly requests Eviction Agent to evict portions of connected clients if there are any;
 * prohibits new connections.

##### Disabling
On disabling, Evacuation Coordinator:
* Disables Eviction Agent.
* Passes to the `disabled` state.

The internal API is:
```erlang
ok = emqx_node_rebalance_evacuation:stop().
```

### Evacuation Coordinator (**with** enforced session evacuation)

The states of this Evacuation Coordinator are: `disabled`, `evicting`, and `takeover`.

![Evacuation Coordinator Statuses](0020-assets/evacuation-coordinator-statuses-enforce.png)

#### Algorithm

##### Starting

At the start, the Evacuation Coordinator reads the settings from the local file.
* If evacuation is enabled Evacuation Coordinator tries to pass to the `evicting` state. It enables the local Eviction Agent.

>     * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passed to the `evicting` state.

Reviewer: ```suggestion
    * If the local Eviction Agent is successfully enabled, then the Evacuation Coordinator passes to the `evicting` state.
```

> # Node evacuation/rebalancing in EMQX 4.x

Reviewer 2: ```suggestion
# Load rebalancing in EMQX 4.x
```
And the file name should be updated too.

## Abstract

> Node rebalancing is the enforced process of client and session migration from a set of nodes to some other ones.

Reviewer 2: ```suggestion
Cluster load rebalancing is the enforced process of client and session migration from a set of nodes to some other ones.
```

Node evacuation is a special case of rebalancing when we want to migrate all connections and sessions from
a particular node.

## Motivation

A user might want to shut down arbitrary nodes, e.g., to maintain the underlying
software or hardware. Or move connections from higher-loaded nodes to lower.

> We need to minimize data loss when a node is switched off. The aim of node evacuation

Reviewer 2: ```suggestion
We need to minimize data loss when a node is switched off (when data persistence to external DB is not in use). The aim of node evacuation
```

Reviewer 2: Nvm, there is this info below

* Facade internal API for
    * connection, session eviction;
    * reporting current connection & session stats.
* HTTP API reporting current status and providing endpoints for health checks.

Eviction Agent has two states: `enabled` and `disabled`.

In the `enabled` state, Eviction Agent has _kind_, a label that indicates the purpose for which the agent is used.
This label is used to avoid a clash between different Rebalance Coordinators, like evacuation and rebalancing.

#### Internal API

Eviction Agent exposes API for the following actions:
* In disabled state:
    * enable the agent;
* In either of enabled states:
    * disable the agent;
    * provide stats about connected/disconnected client count;
    * provide stats about session count;
    * disconnect certain amount of connected clients;
    * evict certain amount of sessions.

#### HTTP API

Eviction Agent exposes two endpoints: `/node_eviction/status` and `/node_eviction/heath_check`.

#### Algorithm

##### `disabled` state

The disabled state is the default. In the `disabled` state Eviction Agent does not influence the node anyhow.

HTTP responses are the following:

Request:
```

> GET /api/v4/node_eviction/status

Reviewer 2: ```suggestion
GET /api/v4/load_rebalance/status
```

```

Response:
```
HTTP/1.1 200 OK

{"status": "disabled"}
```

Request:
```

> GET /api/v4/node_eviction/heath_check

Reviewer 2: Should we consider a more generic api name?
```suggestion
GET /api/v4/load_rebalance/availability_check
```

Author: ✅

```

Response:
```
HTTP/1.1 200 OK
```

The internal API behavior is the following:

```erlang

> disabled = emqx_eviction_agent:status().

Reviewer 2: should we consider renaming this module to a more generic one?
